pkg/globalnet/controllers/ingress_endpoints_controller.go

```diff
@@ -0,0 +1,149 @@
+	}
+
+	fieldSelector := fields.Set(map[string]string{"metadata.name": svc.Name}).AsSelector().String()
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Endpoints syncer",
+		ResourceType:        &corev1.Endpoints{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     svc.Namespace,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		SourceFieldSelector: fieldSelector,
+		ResourcesEquivalent: areEndpointsEqual,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := controller.Start(); err != nil {
+		return nil, err
+	}
+
+	ingressIPs := config.SourceClient.Resource(*gvr).Namespace(corev1.NamespaceAll)
+	controller.reconcile(ingressIPs, func(obj *unstructured.Unstructured) runtime.Object {
+		endpointsName, exists, _ := unstructured.NestedString(obj.Object, "spec", "serviceRef", "name")
+		if exists {
+			return &corev1.Endpoints{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      endpointsName,
+					Namespace: obj.GetNamespace(),
+				},
+			}
+		}
+
+		return nil
+	})
+
+	klog.Infof("Created Endpoints controller for (%s/%s) with selector %q", svc.Namespace, svc.Name, fieldSelector)
+
+	return controller, nil
+}
+
+func (c *ingressEndpointsController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	endpoints := from.(*corev1.Endpoints)
+	key, _ := cache.MetaNamespaceKeyFunc(endpoints)
+
+	ingressIP := &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ep-%.60s", endpoints.Name),
+			Namespace: endpoints.Namespace,
+			Labels: map[string]string{
+				ServiceRefLabel: c.svcName,
+			},
+		},
+	}
+
+	if op == syncer.Delete {
+		c.ingressIPMap.Remove(ingressIP.Name)
+		klog.Infof("Ingress Endpoints %s for service %s deleted", key, c.svcName)
+
+		return ingressIP, false
+	}
+
+	if c.ingressIPMap.Contains(ingressIP.Name) {
+		// Avoid assigning ingressIPs to endpoints that are not ready with an endpoint IP
+		return nil, false
+	}
+
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// TODO: consider handling multiple endpoints IPs
+	if len(endpoints.Subsets) == 0 || len(endpoints.Subsets[0].Addresses) == 0 {
+		return nil, false
+	}
+
+	ingressIP.ObjectMeta.Annotations = map[string]string{
+		headlessSvcEndpointsIP: endpoints.Subsets[0].Addresses[0].IP,
+	}
+
+	ingressIP.Spec = submarinerv1.GlobalIngressIPSpec{
+		Target:     submarinerv1.HeadlessServiceEndpoints,
+		ServiceRef: &corev1.LocalObjectReference{Name: c.svcName},
+	}
+
+	c.ingressIPMap.Add(ingressIP.Name)
+
+	return ingressIP, false
+}
+
+func areEndpointsEqual(obj1, obj2 *unstructured.Unstructured) bool {
+	subsets1, _, _ := unstructured.NestedSlice(obj1.Object, "subsets")
```

Comment:
In my understanding, it needs to be started here, because this controller is started via `IngressEndpointsControllers.start()` per Service basis. It is using the similar approach that ingress_pod_controllers.go does for ingress_pod_controller.go.